Signal/src/Profiles/OWSProfileManager.m

```diff
@@ -199,51 +191,48 @@ - (void)observeNotifications
                                                object:nil];
 }
 
-- (void)appLaunchDidBegin
-{
-    // Do nothing; we only want to make sure this singleton is created on startup.
-}
-
 #pragma mark - User Profile Accessor
 
+// This method can be safely called from any thread.
 - (UserProfile *)getOrCreateUserProfileForRecipientId:(NSString *)recipientId
 {
-    OWSAssert([NSThread isMainThread]);
     OWSAssert(recipientId.length > 0);
 
     __block UserProfile *instance;
     // Make sure to read on the local db connection for consistency.
-    [self.dbConnection readWriteWithBlock:^(YapDatabaseReadWriteTransaction *transaction) {
+    [self.dbConnection readWithBlock:^(YapDatabaseReadTransaction *transaction) {
         instance = [UserProfile fetchObjectWithUniqueID:recipientId transaction:transaction];
-
-        if (!instance) {
-            instance = [[UserProfile alloc] initWithRecipientId:recipientId
-                                                    profileName:nil
-                                                      avatarUrl:nil
-                                                   avatarDigest:nil
-                                                 avatarFileName:nil];
-            [instance saveWithTransaction:transaction];
-        }
     }];
 
+    if (!instance) {
+        instance = [[UserProfile alloc] initWithRecipientId:recipientId];
+    }
+
     OWSAssert(instance);
 
     return instance;
 }
 
-- (nullable UserProfile *)getOrCreateLocalUserProfile
+// All writes to user profiles should occur on the main thread.
+- (void)saveUserProfile:(UserProfile *)userProfile
```

Comment:
Yes, my concern was that mutations have two steps: a) modifying the properties b) writing to database. I didn't want to allow one thread to be modifying a property while another thread was saving the profile, especially since some of our profile properties (like avatar url + digest, and lastUpdateDate + all other properties) should be modifies in lockstep.  

As discussed, I think we can allow writes from any thread if we actually do the writing & saving on a serial queue.  I think it'll work, I just haven't gotten there and haven't committed to doing it yet since I haven't seen a key reason to.